File: apps/minecraft/spigot/src/main/java/crabcraft/net/crabUtilities/jade/protocol/provider/StreamServerDataProvider.java
package crabcraft.net.crabUtilities.jade.protocol.provider;

import net.minecraft.nbt.CompoundTag;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import org.jetbrains.annotations.Nullable;
import crabcraft.net.crabUtilities.jade.protocol.accessor.Accessor;

public interface StreamServerDataProvider<T extends Accessor<?>, D> extends ServerDataProvider<T> {

    @Override
    default void appendServerData(CompoundTag data, T accessor) {
        D value = streamData(accessor);
        if (value != null) {
            data.put(getUid().toString(), accessor.encodeAsNbt(streamCodec(), value));
        }
    }

    @Nullable
    D streamData(T accessor);

    StreamCodec<RegistryFriendlyByteBuf, D> streamCodec();
}

File: apps/minecraft/spigot/src/main/kotlin/crabcraft/net/crabUtilities/jade/protocol/provider/StreamServerDataProvider.kt
package crabcraft.net.crabUtilities.jade.protocol.provider

import crabcraft.net.crabUtilities.jade.protocol.accessor.Accessor
import net.minecraft.nbt.CompoundTag
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.network.codec.StreamCodec

interface StreamServerDataProvider<T : Accessor<*>, D : Any> : ServerDataProvider<T> {
    override fun appendServerData(data: CompoundTag, accessor: T) {
        val value = streamData(accessor) ?: return
        data.put(getUid().toString(), accessor.encodeAsNbt(streamCodec(), value))
    }

    fun streamData(accessor: T): D?

    fun streamCodec(): StreamCodec<RegistryFriendlyByteBuf, D>
}
